### PR TITLE
Remove hardcoded dependency between artifacts and projects

### PR DESCRIPTION
**Pull request by JAORMX**
# Summary

This makes repos nullable in artifacts. As of today, this is not really
a case that actually happens... so this should effectively be a no-op.
However, the idea is that soon we'll start getting artifacts that that
come from other providers that are not the same as the repository they
are linked to.

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

**Review by dmjb — APPROVED** [review]
Approving. It's unfortunate that we need to add the provider name to the artifacts table, but it's a temporary approach so I think it's fine for now.

File: internal/controlplane/handlers_githubwebhooks.go [hunk 1]
```diff
@@ -497,16 +497,22 @@ func (s *Server) parseArtifactPublishedEvent(
...
+		RepositoryID: uuid.NullUUID{
+			UUID:  dbrepo.ID,
+			Valid: true,
+		},
 		ArtifactName:       tempArtifact.GetName(),
 		ArtifactType:       tempArtifact.GetTypeLower(),
 		ArtifactVisibility: tempArtifact.Visibility,
+		ProjectID:          dbrepo.ProjectID,
+		ProviderID:         dbrepo.ProviderID,
+		ProviderName:       dbrepo.Provider,
```
Comment (jhrozek) [inline internal/controlplane/handlers_githubwebhooks.go:509]: I'm getting:
```
{"level":"debug","Timestamp":1712687249228835000,"message":"Error parsing github webhook message: error upserting artifact: pq: there is no unique or exclusion constraint matching the ON CONFLICT specification"}
```
when trying to upsert an artifact with this PR.

My database is on version 50.

File: database/migrations/000050_artifact_projects.up.sql [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+-- make project_id not nullable in artifacts
+ALTER TABLE artifacts ALTER COLUMN project_id SET NOT NULL;
+
+ALTER TABLE artifacts ALTER COLUMN provider_id SET NOT NULL;
+
+ALTER TABLE artifacts ALTER COLUMN provider_name SET NOT NULL;
+
+-- Now that repository_id's are nullable, let's index artifacts by repository_id where the repository_id is not null
+CREATE INDEX artifacts_repository_id_idx ON artifacts (repository_id) WHERE repository_id IS NOT NULL;
```
Comment (jhrozek) [inline database/migrations/000050_artifact_projects.up.sql:65]:
```suggestion
CREATE UNIQUE INDEX artifacts_repository_id_idx ON artifacts (repository_id) WHERE repository_id IS NOT NULL;
```

File: database/migrations/000050_artifact_projects.up.sql [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+ALTER TABLE artifacts ADD COLUMN provider_name TEXT;
+
+-- make provider_id a foreign key to providers
+ALTER TABLE artifacts ADD CONSTRAINT fk_artifacts_provider_id_and_name FOREIGN KEY (provider_id, provider_name) REFERENCES providers (id, name);
+
+-- remove index artifact_name_lower_idx from artifacts
+DROP INDEX artifact_name_lower_idx;
+
+-- recreate index artifact_name_lower_idx on artifacts but with project_id
+CREATE INDEX artifact_name_lower_idx ON artifacts (project_id, LOWER(artifact_name));
```
Comment (jhrozek) [inline database/migrations/000050_artifact_projects.up.sql:36]:
```suggestion
CREATE UNIQUE INDEX artifact_name_lower_idx ON artifacts (project_id, LOWER(artifact_name));
```

**Review by jhrozek — DISMISSED** [review]
this works for me!

**Review by jhrozek — APPROVED** [review]
this works even better for me!

**Pull request merged**